File: day18/p2.py
from collections import deque

size = 70
coords = [list(map(int, line.split(","))) for line in open("input.txt")]

def connected(bytes):
    grid = [[0] * (size + 1) for _ in range(size + 1)]

    for c, r in coords[:bytes]:
        grid[r][c] = 1

    q = deque([(0, 0)])
    seen = {(0, 0)}

    while q:
        r, c = q.popleft()
        for nr, nc in [(r + 1, c), (r, c + 1), (r - 1, c), (r, c - 1)]:
            if nr < 0 or nc < 0 or nr > size or nc > size:
                continue
            if grid[nr][nc] == 1:
                continue
            if (nr, nc) in seen:
                continue
            if nr == nc == size:
                return True
            seen.add((nr, nc))
            q.append((nr, nc))

    return False

low = 0
high = len(coords) - 1

while low < high:
    mid = (low + high) // 2
    if connected(mid + 1):
        low = mid + 1
    else:
        high = mid

print(coords[low])

File: day_18/p2.js
const fs = require('fs');

const size = 70;
const coords = fs.readFileSync('input.txt', 'utf-8').split('\n')
    .map(line => line.split(',').map(Number));

function connected(bytes) {
    const grid = Array.from({ length: size + 1 }, () => Array(size + 1).fill(0));

    // Populate the grid with the coordinates
    coords.slice(0, bytes).forEach(([c, r]) => {
        grid[r][c] = 1;
    });

    const queue = [[0, 0]]; // Start with (0, 0)
    const seen = new Set();
    seen.add('0,0'); // Mark the starting point as visited

    while (queue.length > 0) {
        const [r, c] = queue.shift(); // Dequeue the first element

        const directions = [
            [r + 1, c], [r, c + 1], [r - 1, c], [r, c - 1]
        ];

        for (const [nr, nc] of directions) {
            if (nr < 0 || nc < 0 || nr > size || nc > size) continue;
            if (grid[nr][nc] === 1) continue; // Skip blocked cells

            const key = `${nr},${nc}`;
            if (seen.has(key)) continue; // Skip already visited cells

            if (nr === nc && nr === size) {
                return true; // Return true if we reach the target
            }

            seen.add(key); // Mark as visited
            queue.push([nr, nc]); // Enqueue the new position
        }
    }

    return false; // Return false if no connection is found
}

let low = 0;
let high = coords.length - 1;

while (low < high) {
    const mid = Math.floor((low + high) / 2);
    if (connected(mid + 1)) {
        low = mid + 1;
    } else {
        high = mid;
    }
}

console.log(coords[low]);
